content/post/ddos-detection.md

```diff
@@ -0,0 +1,419 @@
+{"timestamp":1509494400.002,"client":"229.28.98.137","resource":" 594e9919a","server":"224.217.101.60"}
+...
+```
+
+In this example, we will use synthetic data generated with a python script that is included along with the demo code at [ddos-detection].
+The data has the following characteristics:
+
+1. 100 servers, serving 10,000 resources.
+2. 10 seconds of normal traffic, with 1,000 clients making 1,000 requests/second, distributed uniformly over all the servers.
+3. 10 seconds of attack traffic, with 100,000 clients making 25,000 requests/second, with 90% of the traffic hitting only 10 servers.
+4. 10 seconds of normal traffic, with 1,000 clients making 1,000 requests (in total) per second, distributed uniformly over all the servers.
+
+To produce a similar set, run
+
+```bash
+python data_gen.py --clients 1000 --attack-clients 100000 \
+  --requests 1000 --attack-requests 25000 --loaded-weight 0.9 \
+  --file data.json
+```
+
+To consume this stream, we will need to decode the JSON documents with a Wallaroo `Decoder`, using the standard library's `json` module:
+
+```python
+class Decoder(object):
+    def header_length(self):
+        return 4
+
+    def payload_length(self, bs):
+        return struct.unpack(">I", bs)[0]
+
+    def decode(self, bs):
+        return json.loads(bs)
+```
+
+As each server's status is computed independently of the others, we need a way to choose the partition for each incoming record. We could create a partition for each unique IP address, but that would require a partition index list the size of _2^32 - 1_ (assuming only IPv4 addresses), which will be a massive waste of memory given that we only have a small number of servers to monitor (certainly much less than 4.29 billion!). Instead, we will divide the IP range into _K_ even segments to be our partitions. We can then convert the string IP address in each record from the form of `A.B.C.D` to an integer, and determine the index of the segment it is in, which will become its partition identifier:
+
+```python
+class ServerPartitionFunction(object):
+    def __init__(self, partitions):
+        self.part_size = 2**32/partitions
+
+    def partition(self, data):
+        parts = data['server'].split('.')
+        ipnum = ((int(parts[0]) << 24) + (int(parts[1]) << 16) +
+                 (int(parts[2]) << 8) + int(parts[3]))
+        return ipnum/self.part_size
+```
+
+The state object is where things get interesting. For each server, we need to maintain a 60 seconds long rolling window of aggregate data. This data is small enough that we could probably just use a list and pay the cost of shifting it whenever we need to roll the window. But a more efficient way of doing this, especially if the data window is larger, would be to use a [circular buffer]. For brevity, the code for the circular buffer is not included in this post, but you may find in the [demo code][ddos-detection circularbuffer].
+
+Each server's state needs to keep track of a few things:
```

Comment:
"Each server's state"

I don't think that is quite right. Its not the server's state. Its the information we are keeping about the state of a given server.